### ISSUE TITLE
adding normalization at inferece level for applying temp scaling in yolov8n

### ISSUE THREAD

**Issue by omidgh1379** [ultralytics/yolov5#13597]
> Im finetune yolov8n on my own dataset with the goal of getting the raw logits and apply temp scaling to reduce the uncertainty. 
> 
> when I get the raw result from model.model(image) which normalization should I apply to match my traning normalization ? 
> 
> since when I apply without normizaiton I have high overconfidect model and when I divide by 255 my model is underconfident. 
> 
> 
>   def compare_with_original_model(self , valid_loader   ):
> 
>         all_logits = []
>         all_labels =[]
> 
>         
> 
>         with torch.no_grad():
> 
>             self.temperature.data = torch.tensor([1.0] , device = self.temperature.device , dtype = torch.float32)
>             for batch in valid_loader:
>                 img = batch['img']
>                 img = img / 255
>                 
>                 outputs = self.forward(img)
>                 
> 
>                 targets = {
>                     'cls' : batch['cls'] , 
>                     'bboxes' : batch['bboxes'] , 
>                     'batch_idx' : batch['batch_idx']
>                 }
> 
>                 logits , labels = self._match_targets_to_outputs(outputs , targets)
> 
>                 if logits is not None and labels is not None:
>                     all_logits.append(logits)
>                     all_labels.append(labels)
> 
>         all_logits = torch.cat(all_logits , dim = 0)
>         all_labels = torch.cat(all_labels , dim = 0).long()
> 
> 
>         nll_criterion = nn.CrossEntropyLoss().cuda()
>         ece_criterion = _ECE_criterion().cuda()
> 
>         nll = nll_criterion(all_logits , all_labels.squeeze(1)).item()
>         ece = ece_criterion(all_logits , all_labels.squeeze(1)).item()
> 
>         print('all_logits shape is : ',all_logits.shape)
>         print('all_labels shape is : ',all_labels.squeeze(1).shape)
> 
>         print(f'NLL: {nll:.3f}, ECE: {ece:.3f}')
>         print('entropy of the temperature scaled model is : ',self.confidence_entropy(all_logits))
>         return nll , ece 

 _Originally posted by @omidgh1379 in [#12063](https://github.com/ultralytics/yolov5/issues/12063#issuecomment-2888921953)_

**Comment by UltralyticsAssistant** [comment]
👋 Hello @omidgh1379, thank you for reaching out and for your interest in YOLOv5 🚀! This is an automated response to help you get started—an Ultralytics engineer will assist you further soon.

For questions involving inference normalization, temperature scaling, or custom model modifications, please provide a **minimum reproducible example (MRE)** if possible. This helps us understand and debug your scenario more efficiently.

If your issue relates to 🐛 bugs, please share the smallest code snippet and dataset sample that reproduces the problem. For custom training or inference questions, include relevant code, logs, and any clarifying details about your dataset or model setup. You can also review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for guidance.

## Requirements
[**Python>=3.8.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed, including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). Quick install:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```

## Environments
YOLOv5 can be run in these verified environments, all dependencies (including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), [PyTorch](https://pytorch.org/)) preinstalled:
- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

## Status
<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>
If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py) and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your contribution! 🥳

**Comment by pderrenger** [comment]
Hi @omidgh1379,

I notice you're asking about YOLOv8n normalization in the YOLOv5 repository. For consistency between training and inference, you should apply the same normalization in both stages. Since you trained with images normalized by dividing by 255, you should maintain this during inference.

If you're experiencing confidence calibration issues, temperature scaling should be applied to the model's output logits after normalization, not as a replacement for input normalization. The division by 255 is correct for input preprocessing.

For optimal temperature scaling:
1. Keep your input normalization consistent (img = img / 255)
2. Apply temperature scaling to the output logits (dividing logits by your learned temperature parameter)
3. Fine-tune the temperature parameter on a validation set

If you're specifically working with YOLOv8n, you might want to check the [Ultralytics YOLOv8 repository](https://github.com/ultralytics/ultralytics) for architecture-specific details, as there could be slight differences in how normalization is handled.
